### PR TITLE
Support DML predicates in JdbcSqlConnector [HZ-2084]

### PR DESCRIPTION
**Pull request by frant-hartm**
Adds support for predicates in DML (update, delete) for the jdbcSqlConnector.

If an expression is supported is decided by the `SupportsRexVisitor`, which traverses the RexNode to see if all nested expressions are supported.

When the predicate can be pushed down and there is no input, the WriteJdbcP would execute nothing. A fake input source vertex is provided and it produces a single fake row.

The `UpdateProcessorSupplier` and `DeleteProcessorSupplier` were unified into `DmlProcessorSupplier` as they were identical. The dynamic parameters and input refs are collected separately.

New test coverage:
- PredicatePushDownJdbcSqlConnectorTest which tests that expressions from https://docs.hazelcast.com/hazelcast/5.2/sql/functions-and-operators#hide-nav can be pushed down. It runs the actual query and also explain to verify the predicate was pushed down. Runs against H2 (in PR builder) and Postgres + MySQL (nightly).
- UpdatePredicatePushDownJdbcSqlConnectorTest/DeletePredicatePushDownJdbcSqlConnectorTest which test some edge cases for update and delete

Checklist:
- [x] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [x] Label `Add to Release Notes` or `Not Release Notes content` set
- [x] Request reviewers if possible

### CODE REVIEW

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcSqlConnector.java [hunk 1]
```diff
@@ -375,56 +385,117 @@ public Vertex updateProcessor(
...
+        }
+    }
+
+    private static Vertex fakeSourceVertex(DagBuildContext context, String name, String partitionKey) {
+        Vertex v = context.getDag().newUniqueVertex(name,
+                forceTotalParallelismOne(
+                        of(() -> new ConvenientSourceP<>(
+                                ExpressionEvalContext::from,
+                                (evalContext, buf) -> {
+                                    SourceBuffer<Object> buffer = (SourceBuffer<Object>) buf;
```
Comment (k-jamroz) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcSqlConnector.java:485]: to check: how will it behave with processing guarantees? For batch jobs they are not working fully anyway, but would be nice know what will happen in such case

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcSqlConnector.java [hunk 2]
```diff
@@ -375,56 +385,117 @@ public Vertex updateProcessor(
             @Nullable HazelcastRexNode predicate,
             boolean hasInput
     ) {
-        assert predicate == null;
-        assert hasInput;
+        // We don't allow both predicate and input at the same time
+        assert !(predicate != null && hasInput);
```
Comment (k-jamroz) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcSqlConnector.java:389]:
```suggestion
        assert predicate == null || !hasInput;
```

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcSqlConnector.java [hunk 3]
```diff
@@ -375,56 +385,117 @@ public Vertex updateProcessor(
...
-    public Vertex deleteProcessor(@Nonnull DagBuildContext context, @Nullable HazelcastRexNode predicate, boolean hasInput) {
-        // TODO use the predicate
-        assert predicate == null;
+    public Vertex deleteProcessor(
+            @Nonnull DagBuildContext context,
+            @Nullable HazelcastRexNode predicate,
+            boolean hasInput
+    ) {
+        // We don't allow both predicate and input at the same time
+        assert !(predicate != null && hasInput);
```
Comment (k-jamroz) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcSqlConnector.java:423]:
```suggestion
        assert predicate == null || !hasInput;
```

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/jdbc/PredicatePushDownJdbcSqlConnectorTest.java [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+                "CREATE MAPPING " + tableName + " DATA CONNECTION " + TEST_DATABASE_REF
+        );
+    }
+
+    @Test
+    public void runQueryExpectSingleResult_and_FullScanPhysicalRel() {
+        assertRowsAnyOrder(query, new Row("John Doe"));
+
+        assertExplain(
+                "EXPLAIN " + query,
```
Comment (k-jamroz) [inline hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/jdbc/PredicatePushDownJdbcSqlConnectorTest.java:156]: if I understand correctly, this is inteded to check if the plan consist only in FullScanPhysicalRel, without Calc. If so, we can also add explicit check for number of rows in plan = 1

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/SupportsRexVisitor.java [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+            case OTHER:
+            case OTHER_FUNCTION:
+                return OTHER_SUPPORTED.contains(operator.getName());
+
+            default:
+                return false;
+        }
+    }
+
+    private boolean isOperandsSupported(RexCall call) {
```
Comment (k-jamroz) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/SupportsRexVisitor.java:103]: nit: isOperandsSupported -> areOperandsSupported

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/CustomContext.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+import org.apache.calcite.rel.rel2sql.RelToSqlConverter;
+import org.apache.calcite.rel.rel2sql.SqlImplementor;
+import org.apache.calcite.rel.rel2sql.SqlImplementor.Context;
+import org.apache.calcite.sql.SqlDialect;
+import org.apache.calcite.sql.SqlNode;
+
+import java.util.function.IntFunction;
+
+class CustomContext extends Context {
```
Comment (k-jamroz) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/CustomContext.java:27]: can we have more specific name for this class and name or comment explaining its existence?

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcSqlConnector.java [hunk 7]
```diff
@@ -375,56 +385,117 @@ public Vertex updateProcessor(
...
+            // There is no input, we push the whole update query to the database, but we need single fake item
+            // to execute the update in WriteJdbcP
+            // We can consider refactoring the WriteJdbcP, so it doesn't need the fake item in the future.
+
+            // Create random partition key so the fake source and processor run on the same member
+            String partitionKey = UuidUtil.newUnsecureUuidString();
+
+            Vertex v = fakeSourceVertex(context, "FakeSourceFor" + statement, partitionKey);
+            Vertex updateVertex = context.getDag().newUniqueVertex(
+                    statement + "(" + table.getExternalNameList() + ")",
```
Comment (k-jamroz) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcSqlConnector.java:464]: `statement` text in theory can contain literal secrets and be long. Better name would be `update` + table name + columns or `delete` + table name.

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcSqlConnector.java [hunk 8]
```diff
@@ -375,56 +385,117 @@ public Vertex updateProcessor(
...
+
+        if (!hasInput) {
+            // There is no input, we push the whole update query to the database, but we need single fake item
+            // to execute the update in WriteJdbcP
+            // We can consider refactoring the WriteJdbcP, so it doesn't need the fake item in the future.
+
+            // Create random partition key so the fake source and processor run on the same member
+            String partitionKey = UuidUtil.newUnsecureUuidString();
+
+            Vertex v = fakeSourceVertex(context, "FakeSourceFor" + statement, partitionKey);
```
Comment (k-jamroz) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcSqlConnector.java:462]: name of this vertex will be visible for user. `FakeSourceFor` might be puzzling, I'd prefer `DummySourceFor`. Also same comment about the `statement` as below.

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcSqlConnector.java (hunk 7 above)
Comment (k-jamroz) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcSqlConnector.java:464]: ok, statement is only update/delete

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/DMLProcessorSupplier.java [hunk 9]
```diff
@@ -40,34 +40,38 @@
 import static java.util.Collections.singletonList;
 import static java.util.Objects.requireNonNull;
 
-public class UpdateProcessorSupplier
+public class DMLProcessorSupplier
```
Comment (TomaszGaweda) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/DMLProcessorSupplier.java:43]: nit: DmlProcessorSupplier, same as JdbcSqlConnector not JDBCSQLConnector

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/SupportsRexVisitor.java [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+                return OTHER_SUPPORTED.contains(operator.getName());
+
+            default:
+                return false;
+        }
+    }
+
+    private boolean operandsSupported(RexCall call) {
+        for (RexNode operand : call.operands) {
+            if (!operand.accept(this)) {
```
Comment (TomaszGaweda) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/SupportsRexVisitor.java:105]: can it return null?

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/jdbc/UpdatePredicatePushDownJdbcSqlConnectorTest.java [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+
+        assertJdbcRowsAnyOrder(tableName,
+                new Row(0, "updated", 0, JSON),
+                new Row(1, "name-1", 1, JSON)
+        );
+    }
+
+    @Test
+    public void noParameterPredicateCanNotPushDown() throws Exception {
+        execute("UPDATE " + tableName + " SET name = 'updated' WHERE age = 0 AND JSON_QUERY(data, '$.value') = '42'");
```
Comment (TomaszGaweda) [inline hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/jdbc/UpdatePredicatePushDownJdbcSqlConnectorTest.java:74]: Actually here `age = 0` should be pushed down and second predicate should not be, is this the intention of this test? Name suggest that nothing should be pushed down

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcSqlConnector.java [hunk 12]
```diff
@@ -375,56 +385,117 @@ public Vertex updateProcessor(
...
+            Vertex updateVertex = context.getDag().newUniqueVertex(
+                    statement + "(" + table.getExternalNameList() + ")",
+                    forceTotalParallelismOne(processorSupplier, partitionKey)
```
[the remark this answers is not in the document]
Comment (frant-hartm) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcSqlConnector.java:465]: Created HZ-2423 to track this issue.

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/jdbc/PredicatePushDownJdbcSqlConnectorTest.java (hunk 4 above)
Comment (frant-hartm) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/jdbc/PredicatePushDownJdbcSqlConnectorTest.java:157]:
```
            assertThat(rows).containsExactly(rels);
```
here
https://github.com/hazelcast/hazelcast/pull/24521/files#diff-eebb57c32fcda5753e414657a2bc9a3fc615acfae2953f633d4e7c3d773d3fc5R171

checks for the size of `rows` is same as `rels`. Is that enough, or do you want explicit `assertThat(rows).hasSameSizeAs(rels);`?

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/jdbc/UpdatePredicatePushDownJdbcSqlConnectorTest.java (hunk 11 above)
Comment (frant-hartm) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/jdbc/UpdatePredicatePushDownJdbcSqlConnectorTest.java:74]: c2d3f22b390

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/SupportsRexVisitor.java (hunk 10 above)
Comment (frant-hartm) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/SupportsRexVisitor.java]: Yes, good catch!
dbfd7e59b29

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/DMLProcessorSupplier.java (hunk 9 above)
Comment (frant-hartm) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/DMLProcessorSupplier.java]: 85371ae0ddc

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcSqlConnector.java (hunk 8 above)
Comment (frant-hartm) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcSqlConnector.java]: 1ae31f6515a

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/CustomContext.java (hunk 6 above)
Comment (frant-hartm) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/CustomContext.java]: 2bada8f2b25

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcSqlConnector.java (hunk 3 above)
Comment (frant-hartm) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcSqlConnector.java]: ed1365d9a8e

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcSqlConnector.java (hunk 2 above)
Comment (frant-hartm) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcSqlConnector.java]: ed1365d9a8e

**Review by TomaszGaweda — APPROVED** [review]
Well done!

**Review by TomaszGaweda — APPROVED** [review]
Well done!

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcSqlConnector.java [hunk 13]
```diff
@@ -375,56 +385,117 @@ public Vertex updateProcessor(
...
+                    processorSupplier
+            ).localParallelism(1);
+            return updateVertex;
+        }
+    }
+
+    private static Vertex dummySourceVertex(DagBuildContext context, String name, Address localAddress) {
+        Vertex v = context.getDag().newUniqueVertex(name,
+                forceTotalParallelismOne(
+                        of(() -> new ConvenientSourceP<>(
```
Comment (k-jamroz) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcSqlConnector.java:482]: `ConvenientSourceP` is heavy (in particular non-cooperative), do we have something simpler?

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcSqlConnector.java [hunk 14]
```diff
@@ -375,56 +385,117 @@ public Vertex updateProcessor(
...
+                    statement + "(" + table.getExternalNameList() + ")",
+                    forceTotalParallelismOne(processorSupplier, localAddress)
+            );
+
+            context.getDag().edge(Edge.between(v, updateVertex));
+            return updateVertex;
+        } else {
+            Vertex updateVertex = context.getDag().newUniqueVertex(
+                    statement + "(" + table.getExternalNameList() + ")",
+                    processorSupplier
```
Comment (k-jamroz) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcSqlConnector.java:473]: the way in which we create PMS does not obey https://docs.hazelcast.com/hazelcast/5.2/security/native-client-security#connector-permission. See `WriteJdbcP.metaSupplier`. We did not also before but maybe we should?

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcSqlConnector.java (hunk 13 above)
Comment (frant-hartm) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcSqlConnector.java]: b76162084
